### DISCUSSION TITLE
Modern Adorners

### DISCUSSION THREAD

**Discussion by michael-hawker** [CommunityToolkit/Labs-Windows#278] [Proposals]
# Adorners

PR: https://github.com/CommunityToolkit/Labs-Windows/pull/760

## Problem Statement

[Adorners were a big feature in WPF](https://learn.microsoft.com/dotnet/desktop/wpf/controls/adorners-overview) for creative experiences in visual and textual editors to highlight parts of the UI for contextual interaction by a user (among other things).

This experiment is going to investigate how they work in WPF and how we might polyfill that functionality into UWP/WinUI using new components from the Windows Community Toolkit Labs.

Ideally, we have a similar experience and API surface with possible extensions for improvements to tie Adorners more to the XAML layer. At least showing some common patterns/experiences from Adorners done in WPF replicated in UWP/WinUI.

Most-likely out-of-scope is investigating/addressing the obvious gap in [OnRender](https://learn.microsoft.com/dotnet/api/system.windows.uielement.onrender)/[DrawingContext](https://learn.microsoft.com/dotnet/api/system.windows.media.drawingcontext) methods from UIElement compared to WPF.

## Inspiration

Much called out gap from WPF in implementing certain types of experiences.

Implementation of #212 and #213 type controls, provide an easy to use panel that may be of use for this type of overlay system.

There's also the [ScrollViewerExtensions](https://learn.microsoft.com/dotnet/api/microsoft.toolkit.uwp.ui.animations.scrollviewerextensions) for animating/syncing content in a scrollviewer which could be handy?

[XAML Behaviors](https://github.com/Microsoft/XamlBehaviors) and how behaviors work I think will play a role in how a modern Adorner system could work.

## Proposed Solution

![image](https://user-images.githubusercontent.com/24302614/192063441-9b279d71-e941-4f23-baae-7b7d408a6849.png)

### Overall ideas around architecture

There'll be some sort of `AdornerLayer` type component which represents the `Canvas` type space that adorners live on above all other components of an app. I think when the first `Adorner` is attempted to be used, an AdornerLayer is searched for and injected if none found? Otherwise, we have to leave it up to the developer to specify one within their app (which I think allowing and targeting specific layers would be good to have anyway).

`Adorner` can be subclassed by an implementor for specific 'behaviors' that can be shared. However, it'd be nice if there was a way to just inject XAML to be shown/placed on the adorner layer without having to write specific implementations using an `Adorner` class. In this case it'd be nice if there was an easy way to toggle the `Adorner` on/off, maybe that's just visibility?

`Adorner`s should know where their attached component is in relation to the display surface (i.e. able to use relative coordinates for layout).

### API Surface

Been using a `labs:AdornerLayer.Xaml` attached property helper at the moment, but expect that to be temporary and turn into it's own Adorner helper that is shipped with this feature.

Expect the actual API surface to be something like:

```xml
 <Button>
    <SymbolIcon HorizontalAlignment="Center"
                Symbol="Mail" />

    <labs:AdornerLayer.Adorner>
      <labs:XamlAdorner>
        <muxc:InfoBadge Margin="-4"
                        HorizontalAlignment="Right"
                        VerticalAlignment="Top"
                        IsHitTestVisible="False"
                        Value="5" />
      </labs:XamlAdorner>
    </labs:AdornerLayer.Xaml>
  </Button>
```

We want an `AdornerDecorator` helper to help establish a way for a developer to inject their own AdornerLayers (especially at the top-level due to issues with RootScrollViewer (TODO Link to WinUI issue to be filed)):

```
<AdornerDecorator>
  <ContentPresenter/><!-- or like app's root Frame -->
</AdornerDecorator>
```
From https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/Themes/XAML/Window.xaml

## API Reference Docs
- [Adorner](https://learn.microsoft.com/dotnet/desktop/wpf/controls/adorners-overview)
- [AdornerLayer Implementation (WPF)](https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Documents/AdornerLayer.cs)
- [Adorner Implementation (WPF)](https://github.com/dotnet/wpf/blob/main/src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Documents/Adorner.cs)

## References
- [InfoBadge proposal notes](https://github.com/microsoft/microsoft-ui-xaml/issues/2963)

## Platform Issues
- https://github.com/microsoft/microsoft-ui-xaml/issues/5530
- https://github.com/unoplatform/uno/issues/7893

**Comment by michael-hawker** [comment]
Tinkering with ideas for XAML based attaching/initializing of adorners.

Obviously, it'd be one of these types of things, but just iterating on different ideas together here:

```xml
        <Button labs:Adorner.Attach="local:MyAdornerType"> <!-- auto-instantiate by type, reflection is meh -->
            <SymbolIcon Symbol="Mail" HorizontalAlignment="Center"/>
            <labs:Adorner.Xaml> <!-- Simple helper to add xaml directly as an adornment, but then how do we control position, etc...? -->
                <muxc:InfoBadge Value="5"/>
            </labs:Adorner.Xaml>
            <labs:Adorner.Attach> <!-- More generalized point -->
                <local:MyAdornerType/> <!-- Take initialized by xaml version, would need property to get element from attach helper? -->
                <labs:XamlAdorner Offset="??"> <!-- Then we could have our own XAML based adorner to helper here instead which could have more options... -->
                    <muxc:InfoBadge Value="5"/>
                </labs:XamlAdorner>
            </labs:Adorner.Attach>
        </Button>
```

**Comment by michael-hawker** [comment]
Started a branch here: https://github.com/michael-hawker/Labs-Windows/tree/llama/adorners/labs/Adorners

Had a lot of trouble trying to auto-inject a AdornerLayer and re-parent the content of the app scroll viewer. Kept getting the child of another element exception even after waiting for content to be set to null or content unloaded event to fire... 🙁

So easy to add a badge to a button though now with this initial exploration:

```xml
    <Grid>
            <Button>
                <SymbolIcon HorizontalAlignment="Center"
                            Symbol="Mail" />

                <labs:AdornerLayer.Xaml>
                    <muxc:InfoBadge Margin="-4"
                                    HorizontalAlignment="Right"
                                    VerticalAlignment="Top"
                                    IsHitTestVisible="False"
                                    Value="5" />
                </labs:AdornerLayer.Xaml>

            </Button>

        <!-- Dev should have this as part of their main host window/page content above the frame -->
        <labs:AdornerLayer />
    </Grid>
```

Even stumbled into the `GetAdornerLayer` static method on the `AdornerLayer` class, so that part is the same at the moment for WPF, just started by having this simple `Xaml` connector as an attached property, may change in the future, but wanted to get something to start testing. Just got hung up on the auto-insertion of the layer.

Still many things to investigate and test around scrolling, resizing, and visibility of the adorned element. But it's a start!

**Reply by michael-hawker** [reply]
Still having trouble with auto-insertion of a `AdornerLayer`, but have some new thoughts in this space. Did check with WPF that it handles scrolling perfectly and maintains connection to the element during scrolling, so do need to figure this out. Will be the next project...

In the meantime, have lots of notes and potential helpers along with cleaning up the original sample for comparison with the InfoBadge WinUI sample.

**Reply by michael-hawker** [reply]
Turns out my issues were specifically with `RootScrollViewer` at the top of the application, it looks like a `ScrollViewer` but behaves differently. If I try and inject in a normal `ScrollViewer` that's been added by the app developer, than it works fine!

Added another sample, bit more work left to do on that one, but making progress.

Main to dos left for now:
- [ ] investigate RootScrollViewer solution
- [x] `AdornerDecorator` Container (optimization)
  - [ ] `AdornerDecoratorPanel` (optimization) simple `Grid` like panel where all children fill up the available space
- [ ] Detect adorned element size/position change (e.g. ScrollViewer's inner panel layout update event?)
- [ ] Samples
  - [x] Notification Badge w/ InfoBadge
  - [x] Highlight element bounds (partially implemented need toolkit helper port or abstraction)
  - [ ] Resize Grippers
- [ ] Custom adorners i.e. `Adorner` class itself
- [ ] Tests

**Reply by michael-hawker** [reply]
Quick note here to not forget that there was a clipping option for adorners, and I think with our clip helper in the Toolkit, this might not be too big a deal to implement: https://learn.microsoft.com/en-us/dotnet/api/system.windows.documents.adorner.isclipenabled

**Reply by michael-hawker** [reply]
Updated comments and main issue with latest thoughts. Added `AdornerDecorator` polyfill as well now to match WPF and help us with detecting already injected `AdornerLayout` positions.

**Reply by wbokkers** [reply]
What's the state of the Adorner implementation? Any change it will eventually land into the toolkit?

**Reply by michael-hawker** [reply]
I should have probably just merged what I had into Labs here. Our tooling has updated quite a bit, so not sure how extensive rebasing it would be.

There were a number of edge scenarios that needed investigation/work, but I think the base concept was potential viable path. I just also wanted to do some comparison to WPF as a possible polyfill/migration path, but not as familiar as I once was with that side of things.

If anyone wants to help move any of this along, the code should be all in my branch there.

**Reply by Arlodotexe** [reply]
Would be good to at least get it open as a draft, along with some info on the pending edge scenarios that needs work.

**Reply by michael-hawker** [reply]
Yeah, this is a good skeleton to dust off for sure. I'm hoping things lighten up a bit in December, so we should bubble this one up. Would be good to find a WPF Adorner expert to help us vet the migration path or scenarios and such.

**Comment by SongOfYouth** [comment]
any update for this?

**Reply by michael-hawker** [reply]
Indeed! I think I finally figured out a smart way, just by construction, to deal with layout changes of elements.

I need to reconcile a fork of my old branch that I had with this updated test code, but I'll open a PR soon to bring this officially into Labs. That was really just the missing link, and I think I have that figured out now. It's also a good excuse to write a proper `Adorner` base class (optional with new XAML based approach, but also will make more complex scenarios easier).

I can also write up some of the notes from differences with WPF, as I don't remember if I put that in the docs I had in my old branch.

**Comment by duckblaster** [comment]
Any chance of a rotation adorner to go with the resize adorner? And/or a sample with a zoomable Canvas?
